dlib/serialize.h

```diff
@@ -933,6 +1092,88 @@ namespace dlib
 
 // ----------------------------------------------------------------------------------------
 
+    template<std::size_t I = 0, typename FuncT, typename... Tp>
+    inline typename std::enable_if<I == sizeof...(Tp), void>::type
+    for_each_in_tuple(std::tuple<Tp...>&, FuncT)
+    {}
+
+    template<std::size_t I = 0, typename FuncT, typename... Tp>
+    inline typename std::enable_if<I < sizeof...(Tp), void>::type
+    for_each_in_tuple(std::tuple<Tp...>& t, FuncT f)
+    {
+        f(std::get<I>(t));
+        for_each_in_tuple<I + 1, FuncT, Tp...>(t, f);
+    }
+    
+    template<std::size_t I = 0, typename FuncT, typename... Tp>
+    inline typename std::enable_if<I == sizeof...(Tp), void>::type
+    for_each_in_tuple(const std::tuple<Tp...>&, FuncT)
+    {}
+
+    template<std::size_t I = 0, typename FuncT, typename... Tp>
+    inline typename std::enable_if<I < sizeof...(Tp), void>::type
+    for_each_in_tuple(const std::tuple<Tp...>& t, FuncT f)
+    {
+        f(std::get<I>(t));
+        for_each_in_tuple<I + 1, FuncT, Tp...>(t, f);
+    }
+    
+    struct serialize_tuple_helper
+    {
+        serialize_tuple_helper(std::ostream& out_) : out(out_) {}
+        
+        template<typename T>
+        void operator()(const T& item)
+        {
+            serialize(item, out);
+        }
+                
+        std::ostream& out;
+    };
+    
+    struct deserialize_tuple_helper
+    {
+        deserialize_tuple_helper(std::istream& in_) : in(in_) {}
+        
+        template<typename T>
+        void operator()(T& item)
+        {
+            deserialize(item, in);
+        }
+                
+        std::istream& in;
+    };
```

Comment:
Use lambdas instead of writing out function classes.

Reply:
Can I use lambdas here in C++11? C++11 doesn't support generic lambdas. I must be missing something here. Can you be more specific?

Reply:
Oh right, yeah. Forgot that’s c++14 :(